mountpoint-s3-fs/src/logging/event_log.rs

```diff
@@ -0,0 +1,224 @@
+use std::fs::File;
+use std::io::BufWriter;
+use std::path::Path;
+
+/// Provides callback ([EventLogger::log_error_callback]) for logging events in a structured format to a file.
+///
+/// The output format is `\n`-separated `json`'s, where each `json` describes a single event. Currently,
+/// only errors occurring on fuse operations are logged as events. For some failed fuse operations an event
+/// will be present in the log. On contrary, errors occurring before the mount won't be logged (yet).
+///
+/// Fields `error_code`, `s3_error_http_status` and `s3_error_code` may be used to detect specific
+/// failure conditions, for instance, errors caused by throttling or a lack of permissions. Fields
+/// `s3_object_key` and `s3_error_message` may be used to provide further diagnostics to the user.
+///
+/// Note that most of the fields are optional and may not be present in the event. Only `operation`,
+/// `error_code`, `timestamp` amd `version` will be present in all events. The field `error_code`
+/// is assigned to errors by Mountpoint itself and may be used for a rough classification of those.
+/// See [mountpoint_s3_fs::fs::error_metadata] for the list of possible values.
+///
+/// As an example, the following rule may be used to detect permission errors:
+/// `s3_error_http_status == 403 || (s3_error_http_status == 400 && s3_error_code in {"AccessDenied"})`
+///
+/// And the following rule for throttling errors:
+/// `s3_error_http_status == 503`
+use crate::fs::error_metadata::{MOUNTPOINT_ERROR_INTERNAL, MOUNTPOINT_ERROR_LOOKUP_NONEXISTENT};
+use crate::fs::Error;
+use crate::sync::mpsc::{self, Receiver, SyncSender};
+use serde::{Deserialize, Serialize};
+use std::thread::{self, JoinHandle};
+use time::{serde::rfc3339, OffsetDateTime};
+
+const EVENT_LOG_FILE_NAME: &str = "event_log";
+const VERSION: &str = "1";
+
+pub struct EventLogger {
+    events_sender: Option<SyncSender<Event>>,
+    writer_thread: Option<JoinHandle<()>>,
+}
+
+impl EventLogger {
+    pub fn new<P: AsRef<Path>>(log_directory: P) -> anyhow::Result<Self> {
+        let max_inflight_events = 1000;
+        let (tx, rx) = mpsc::sync_channel(max_inflight_events);
+        let file = File::create(log_directory.as_ref().join(EVENT_LOG_FILE_NAME))?;
+        let writer_thread = thread::spawn(|| Self::write_to_file(rx, file));
+        Ok(Self {
+            events_sender: Some(tx),
+            writer_thread: Some(writer_thread),
+        })
+    }
+
+    pub fn log_error_callback(&self) -> crate::fuse::ErrorCallback {
```

Comment:
Why return a new callback? Shouldn't we just return the original on construction? We would not even need a multi-producer channel.